arch/arm/arm-m/systick/systick.c

```diff
@@ -81,38 +86,48 @@ status_t platform_set_periodic_timer(platform_timer_callback callback, void *arg
     return NO_ERROR;
 }
 
+/* Helper to atomically get the global ticks and the current counter value */
+static void systick_get_ticks_val(uint64_t *ticks, uint32_t *val) {
+    spin_lock_saved_state_t state;
+    arch_interrupt_save(&state, SPIN_LOCK_FLAG_INTERRUPTS);
+
+    *val = SysTick->VAL;
+    if (unlikely(SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk)) {
+        current_ticks++;
```

Comment:
Isn't there an issue here if this routine is called twice in a row while interrupts are still disabled? It seems like it'd increment the current_ticks twice. Seems like it'd work cleaner if it just returned current_ticks + 1 in this case and still rely on an interrupt getting a chance to fire within the defined interval.

Reply:
The COUNTFLAG bit will only be observed high once per wrap around, so I think this is functionally correct.

From the ARMv7-M Arrchitecture Reference Manual:

> When the counter transitions to zero, it sets the COUNTFLAG status bit to 1. Reading the COUNTFLAG status bit clears it to 0.

The v6m and v8m manuals contain equivalent language. 

Incrementing the counter directly in this case ensures that time remains monotonic even if we are unable to take the systick exception for more than one wraparound period.

Reply:
Ahhhh clear on read. That makes sense. I generally hate clear on read bits which is why I hadn't considered that they had done it but since they had it does seem to clean up this problem. Also since it's implemented all the way back through v6m (lots of cortex-m0s floating around that LK runs on) it works for all cores.

Reply:
Yea the clear on read was surprising to learn about and somewhat annoying, but it works.

I also feel like they missed an opportunity with the placment of the bit. It's in a separate CSR from the actual count value, and we have to consider the possibility of the counter wrapping in between the two separate CSR reads, and do another read if we see the flag set to ensure the value is from after the wrap around. But... the value only occupies the low 24 bits of the readback CSR. They could have just stuck the count flag in one of the upper bits and made it one nice clean atomic read. Oh well.